accepted/cranelift-roadmap-2022.md

```diff
@@ -0,0 +1,895 @@
+knobs?
+
+The intuition is that, the closer we have such a free-to-turn knob to
+the compiler state that can trigger a bug, the less of the compiler
+the fuzzing engine has to reverse-engineer to find a suitable
+bug-triggering input. Said another way: fuzzers are great at finding
+inverse functions but it's still costly; better to remove the
+indirection in order to multiply their power per unit of time spent.
+
+In addition, not all of the state space in intermediate portions of
+the pipeline may be reachable, or easily reachable, by choosing some
+input program. Perhaps an earlier compiler pass has some normalization
+effect that masks attempts at exploring the space. Or perhaps it is
+just very costly (e.g.: a long sequence of operations that causes a
+particular spill pattern) to reach some of the configurations we want
+to test.
+
+So, seen in that way, providing "chaos mode" instrumentation
+throughout the pipeline should give us a unit-testing-like
+factorization that allows bugs in each stage of the compiler to be
+independently exposed.
+
+The list of potential knobs to turn is long but a few that are likely
+to be reasonably easy to implement are:
+
+* Randomly spill or reload a value in the register allocator.
+* Randomly clobber some state that should be undefined: high bits of a
+  register, or an unused spill-slot or register, or caller-saves at
+  the start of a prologue, or callee-saves at callsites.
+* Randomly reorder code.
+* Randomly skip merging loads with ALU operations, but do the load
+  again at the ALU operation and compare.
+* Randomly choose different instruction lowerings (see below!).
+* Randomly disable certain optimizations, independently at each
+  potential optimization site. Applicable in the middle-end for each
+  optimization pass (e.g.: randomly do not hoist a particular loop
+  statement in LICM), and in backend infrastructure, e.g. randomly do
+  not let instruction selection see further up an operand subtree.
+  
+Some of these transforms (checking ALU op + load consistency,
+clobbering state that should be unusable) can be enabled
+unconditionally as well; they then fit into a sort of "address
+sanitizer" or "debug asserts" sort of philosophy, where extra checks
+are always performed.
+
+Finally, note that there are two very interesting uses beyond the
+straightforward "fuzz with perturbations". First, *statistically sound
+benchmarking* (as implemented by Sightglass) might be able to take
+advantage of, e.g., measuring a set of runs that have been compiled
+with random function or basic-block orderings. This will need some
+careful consideration to separate what it is we wish to measure from
+what we wish to control for; but in principle, having the capability
+of randomizing small bits of the compiler pipeline at will is very
+powerful.
+
+Second, random perturbations can actually be seen as a special case of
+a general meta-framework or "control plane" for the compiler that
```

Comment:
A moonshot idea, but what about exposing these details to the user. Then they could try to make better decisions than the compiler while still being guaranteed (assuming no bugs in clif) that the output isn't a miscompilation.

Reply:
This is definitely way-out-there, but I do like the idea of a programmable meta layer as a first-class user-exposed interface. In the EDA (electronic design automation) world, tools have pretty much all standardized on Tcl as a command language, and one can write scripts like "optimization1; optimization2; simplify x y z; ...". Similarly for the "strategy layer" in proof-automation/verification tools. This would need a lot of thought and is lower-priority than any of the work that directly benefits correctness testing, but, yes, cool idea!